File: src/main/java/cube8540/oauth/authentication/credentials/oauth/token/application/RefreshTokenGranter.java
package cube8540.oauth.authentication.credentials.oauth.token.application;

import cube8540.oauth.authentication.credentials.AuthorityCode;
import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2ClientId;
import cube8540.oauth.authentication.credentials.oauth.error.InvalidClientException;
import cube8540.oauth.authentication.credentials.oauth.error.InvalidGrantException;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetails;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2TokenRequest;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AccessTokenRepository;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AuthorizedAccessToken;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AuthorizedRefreshToken;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2RefreshTokenRepository;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2TokenId;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2TokenIdGenerator;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class RefreshTokenGranter extends AbstractOAuth2TokenGranter {

    private final OAuth2RefreshTokenRepository refreshTokenRepository;

    public RefreshTokenGranter(OAuth2AccessTokenRepository tokenRepository,
                               OAuth2RefreshTokenRepository refreshTokenRepository,
                               @Qualifier("defaultTokenIdGenerator") OAuth2TokenIdGenerator tokenIdGenerator) {
        super(tokenIdGenerator, tokenRepository);
        this.refreshTokenRepository = refreshTokenRepository;
    }

    @Override
    public OAuth2AuthorizedAccessToken createAccessToken(OAuth2ClientDetails clientDetails, OAuth2TokenRequest tokenRequest) {
        OAuth2AuthorizedRefreshToken storedRefreshToken = refreshTokenRepository.findById(new OAuth2TokenId(tokenRequest.getRefreshToken()))
                .orElseThrow(() -> InvalidGrantException.invalidGrant("invalid refresh token"));
        OAuth2AuthorizedAccessToken storedAccessToken = storedRefreshToken.getAccessToken();
        if (!storedAccessToken.getClient().equals(new OAuth2ClientId(clientDetails.getClientId()))) {
            throw InvalidClientException.invalidClient("invalid refresh token");
        }

        refreshTokenRepository.delete(storedRefreshToken);
        if (storedRefreshToken.isExpired()) {
            throw InvalidGrantException.invalidGrant("refresh token is expired");
        }

        Set<String> storedAccessTokenScopes = storedAccessToken.getScopes().stream().map(AuthorityCode::getValue).collect(Collectors.toSet());
        if (!getTokenRequestValidator().validateScopes(storedAccessTokenScopes, tokenRequest.getScopes())) {
            throw InvalidGrantException.invalidScope("cannot grant scope");
        }
        OAuth2AuthorizedAccessToken accessToken = OAuth2AuthorizedAccessToken.builder(getTokenIdGenerator())
                .expiration(extractTokenExpiration(clientDetails))
                .client(storedAccessToken.getClient())
                .username(storedAccessToken.getUsername())
                .scopes(extractGrantScope(storedAccessToken, tokenRequest))
                .tokenGrantType(storedAccessToken.getTokenGrantType())
                .issuedAt(LocalDateTime.now(clock))
                .build();
        accessToken.generateRefreshToken(refreshTokenGenerator(), extractRefreshTokenExpiration(clientDetails));
        accessToken.generateComposeUniqueKey(getComposeUniqueKeyGenerator());
        return accessToken;
    }

    protected Set<AuthorityCode> extractGrantScope(OAuth2AuthorizedAccessToken accessToken, OAuth2TokenRequest tokenRequest) {
        return (tokenRequest.getScopes() == null || tokenRequest.getScopes().isEmpty()) ? accessToken.getScopes() :
                tokenRequest.getScopes().stream().map(AuthorityCode::new).collect(Collectors.toSet());
    }

    @Override
    protected boolean isReturnsExistsToken(OAuth2AuthorizedAccessToken existsAccessToken, OAuth2AuthorizedAccessToken newAccessToken) {
        return false;
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/credentials/oauth/token/application/RefreshTokenGranter.kt
package cube8540.oauth.authentication.credentials.oauth.token.application

import cube8540.oauth.authentication.credentials.AuthorityCode
import cube8540.oauth.authentication.credentials.oauth.client.domain.OAuth2ClientId
import cube8540.oauth.authentication.credentials.oauth.error.InvalidClientException
import cube8540.oauth.authentication.credentials.oauth.error.InvalidGrantException
import cube8540.oauth.authentication.credentials.oauth.error.InvalidRequestException
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetails
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2TokenRequest
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AccessTokenRepository
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AuthorizedAccessToken
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2RefreshTokenRepository
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2TokenId
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2TokenIdGenerator
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class RefreshTokenGranter @Autowired constructor(
    tokenRepository: OAuth2AccessTokenRepository,

    @Qualifier("OAuth2RefreshTokenRepository")
    private val refreshTokenRepository: OAuth2RefreshTokenRepository,

    @Qualifier("defaultTokenIdGenerator")
    tokenIdGenerator: OAuth2TokenIdGenerator
): AbstractOAuth2TokenGranter(tokenIdGenerator, tokenRepository) {

    override fun createAccessToken(clientDetails: OAuth2ClientDetails, tokenRequest: OAuth2TokenRequest): OAuth2AuthorizedAccessToken {
        if (tokenRequest.refreshToken == null) {
            throw InvalidRequestException.invalidRequest("refresh token is required")
        }
        val storedRefreshToken = refreshTokenRepository.findById(OAuth2TokenId(tokenRequest.refreshToken!!))
            .orElseThrow { InvalidGrantException.invalidGrant("invalid refresh token") }

        val storedAccessToken = storedRefreshToken.accessToken
        if (storedAccessToken.client != OAuth2ClientId(clientDetails.clientId)) {
            throw InvalidClientException.invalidClient("invalid refresh token")
        }

        refreshTokenRepository.delete(storedRefreshToken)
        if (storedRefreshToken.isExpired()) {
            throw InvalidGrantException.invalidGrant("refresh token is expired")
        }

        val storedAccessTokenScopes = storedAccessToken.scopes.map(AuthorityCode::value).toSet()
        if (!tokenRequestValidator.validateScopes(storedAccessTokenScopes, tokenRequest.scopes)) {
            throw InvalidGrantException.invalidScope("cannot grant scope")
        }

        val accessToken = makeNewToken(storedAccessToken, clientDetails, tokenRequest)
        accessToken.generateRefreshToken(refreshTokenGenerator(), extractRefreshTokenExpiration(clientDetails))
        accessToken.generateComposeUniqueKey(composeUniqueKeyGenerator)
        return accessToken
    }

    protected fun extractGrantScope(accessToken: OAuth2AuthorizedAccessToken, tokenRequest: OAuth2TokenRequest): Set<AuthorityCode> {
        return if (tokenRequest.scopes != null && tokenRequest.scopes!!.isNotEmpty()) {
            tokenRequest.scopes!!.map { code -> AuthorityCode(code) }.toSet()
        } else {
            accessToken.scopes.toSet()
        }
    }

    override fun isReturnsExistsToken(existsAccessToken: OAuth2AuthorizedAccessToken, newAccessToken: OAuth2AuthorizedAccessToken): Boolean = false

    private fun makeNewToken(storedAccessToken: OAuth2AuthorizedAccessToken, clientDetails: OAuth2ClientDetails, tokenRequest: OAuth2TokenRequest) =
        OAuth2AuthorizedAccessToken(
            tokenIdGenerator = tokenIdGenerator,
            username = storedAccessToken.username,
            client = storedAccessToken.client,
            scopes = extractGrantScope(storedAccessToken, tokenRequest).toMutableSet(),
            tokenGrantType = storedAccessToken.tokenGrantType,
            expiration = extractTokenExpiration(clientDetails),
            issuedAt = LocalDateTime.now(clock)
        )

}